mgr/volumes: subvolumegroup quotas

kotreshhr opened the pull request:
Fixes: https://tracker.ceph.com/issues/53509
Signed-off-by: Kotresh HR <khiremat@redhat.com>

<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".

  - The Signed-off-by line in every git commit is important; see <span class="x x-first x-last">[Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/master/</span>SubmittingPatches.rst<span class="x x-first x-last">)</span>
-->

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

kotreshhr commented:
Introducing subvolumegroup quotas hits this known issue [1]. The issue is hit while removing subvolume under the subvolume group (whose quota is set) as removing subvolumes renames them out of the subvolume group directory (quota root directory). We need to think about solving the issue for this use case.

[1] https://tracker.ceph.com/issues/16884

kotreshhr commented:
> Introducing subvolumegroup quotas hits this known issue [1]. The issue is hit while removing subvolume under the subvolume group (whose quota is set) as removing subvolumes renames them out of the subvolume group directory (quota root directory). We need to think about solving the issue for this use case.
> 
> [1] https://tracker.ceph.com/issues/16884

The issue is addressed using the existing per subvolume trash framework (introduced for subvolume removal with retain snapshot usecase). The commit https://github.com/ceph/ceph/pull/44347/commits/420541369cb7c09becc4c6fde7d7d2f1b76e60f4 in this PR addresses the issue.

vshankar commented:
> The issue is addressed using the existing per subvolume trash framework (introduced for subvolume removal with retain snapshot usecase). The commit [4205413](https://github.com/ceph/ceph/commit/420541369cb7c09becc4c6fde7d7d2f1b76e60f4) in this PR addresses the issue.

What about older implementation that does not have per-subvolume trash? Will this get backported to that release?

kotreshhr commented:
> > The issue is addressed using the existing per subvolume trash framework (introduced for subvolume removal with retain snapshot usecase). The commit [4205413](https://github.com/ceph/ceph/commit/420541369cb7c09becc4c6fde7d7d2f1b76e60f4) in this PR addresses the issue.
> 
> What about older implementation that does not have per-subvolume trash? Will this get backported to that release?

I have moved the relevant subvolume trash definitions to subvolume V1 calls. It handles the subvolume removal on legacy subvolumes if quota is set on group. PTAL

kotreshhr commented:
jenkins test api

kotreshhr commented:
jenkins test api

ajarr commented on src/pybind/mgr/volumes/fs/operations/versions/subvolume_v2.py:
```diff
@@ -370,13 +341,17 @@ def remove(self, retainsnaps=False, internal_cleanup=False):
-            # Delete the volume meta file, if it's not already deleted
-            self.auth_mdata_mgr.delete_subvolume_metadata_file(self.group.groupname, self.subvolname)
+            self.trash_incarnation_dir(retainsnaps)
+            # Update metadata only for retainsnaps, don't update metadata on suvolume
```
I'm unable to follow this comment. 
"don't update metadata on suvolume removal with group quota set, it fails when the group quota is exceeded."

So if a subvolume group has exceeded its quota, then the subvolume within the group cannot be removed ?

kotreshhr replied:
PTAL and mark it resolved, if the rephrase looks fine.

kotreshhr replied:
Good catch! I will test this out. I think this would fail. The correct design is to keep the internal metadata files out of the quota radar but unfortunately `.meta` file resides inside subvolume directory. I will think over it and update what can be done.

kotreshhr replied:
Updating discussion happened on IRC here.

Yes, this would fail. If the group quota is exceeded, removing a subvolume with retainsnapshot fails. Any subvolume operation which updates the metadata would fail. 

The following options were considered to resolve this
1. Pre-allocate the subvolume metadata file:
             Didn't work. Python configparser failed to parse the pre-allocated config file. The config file is preallocated with default args, where the inode size is updated and bytes are filled with zeros. Updating inode size is required to avoid enforcing quota.
             Also, even if it works, pre-allocation needs to be done for all the existing subvolumes while setting quota on subvolumegroup which might not scale if there are thousands of subvolumes.
             
 2. Don't enforce quota if it's mgr client:
              The quota is enforced on the client and ceph doesn't distinguish between internal and external clients.
 
 3. Re-use/Over load existing FULL caps:
              The FULL caps, currently allow the write even if the osds are full if the client holds it. We can overload the same to not enforce quota as well. @gregsfortytwo thinks it's messy?

vshankar replied:
- option 1 would surely give headaches (even if we can make it work)
- option 2 is probably alright, but Option 3 seems like the cleanest of the "hacks'.

kotreshhr replied:
I was exploring further on this, thought of adding client option to disable quota. While doing so, stumbled on the commit https://github.com/ceph/ceph/pull/14978/commits/da9f1d829484fe35e6d5839c19afa371e444c194 which made the quota non-optional. If we can make quota optional again, and disable it for`mgr` client, a lot of issues which pop up because of setting quota on subvolumegroup goes away as discussed below.

1. We no longer need the commit https://github.com/ceph/ceph/commit/420541369cb7c09becc4c6fde7d7d2f1b76e60f4
    which forces to keep the subvolume trash directory within subvolumegroup.
2. Because of 1, subvolumegroup deletion return EAGAIN until the `trash` directory is empty.

@vshankar @batrick @ajarr @gregsfortytwo What do you guys think ?

vshankar replied:
If this special option can be _only_ for the libcephfs instance in ceph-mgr, that would be neat.

Maybe introduce a ceph user for the libcephfs instance in ceph-mgr?

kotreshhr replied:
Updated the PR with the option to disable quota bytes enforcement. The option is disabled for mgr libcephfs connection.

Observations:
1. With this PR, the quota enforcement is disabled for all mgr libcephfs connections but doesn't affect other clients by default.
2. Doing `ceph config set client client_quota_bytes false` would affect all the fuse mounts.

kotreshhr replied:
@vshankar @ceph/cephfs Observations (contd..)
3. The client config set can be done by the users who has `write` caps on the `mon` i.e `mon allow rw`. The same is applicable to LibcephFS connections as well. But the LibcephFS connection can also be initialized via rados object as done in the `mgr` module. In `mgr` module the `rados` object is initialized via a specially constructed context (https://docs.ceph.com/en/latest/mgr/modules/#accessing-rados-or-cephfs). I believe it's applicable even for the LibCephFS connection which is initailized via rados object ? 
4. A user with `mon allow r` caps can mount the CephFS volume. So I believe giving out `mon allow rw` is giving out part of `admin` access which is not done unless required.

So I believe we are good to make quota optional and only the user with `mon allow rw` can enable/disable quota?

ajarr reviewed (commented):
https://github.com/ceph/ceph/pull/44347/commits/2a7ac768d37492924dc159a283ed69610bb38179  is the tricky one. So does this commit fix the issue https://tracker.ceph.com/issues/16884 for legacy subvolumes?

kotreshhr commented on src/pybind/mgr/volumes/fs/operations/group.py:
```diff
@@ -128,6 +128,52 @@ def info(self):
+            if not (newsize == "inf" or newsize == "infinite"):
+                raise (VolumeException(-errno.EINVAL,
+                                       "invalid size option '{0}'"
+                                       .format(newsize)))
```
Nope, I will get it fixed.

kotreshhr commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -824,6 +824,338 @@ def test_subvolume_group_info(self):
+        usedsize = int(self.mount_a.getfattr(subvolpath, "ceph.dir.rbytes"))
+        susedsize = int(self.mount_a.run_shell(['stat', '-c' '%s', subvolpath]).stdout.getvalue().strip())
+        if isinstance(self.mount_a, FuseMount):
+            # kclient dir does not have size==rbytes
```
I remember trying it on kclient and it works. This commit 89a18171342e from @batrick added this. May be there is some bug due to which size doesn't match upto rbytes? I just chose to retain that check. @batrick Any idea ?

kotreshhr replied:
I think kclient can fetch ceph.dir.rbytes but size from stat would not be match?

kotreshhr commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -824,6 +824,338 @@ def test_subvolume_group_info(self):
+        susedsize = int(self.mount_a.run_shell(['stat', '-c' '%s', subvolpath]).stdout.getvalue().strip())
+        if isinstance(self.mount_a, FuseMount):
+            # kclient dir does not have size==rbytes
+            self.assertEqual(usedsize, susedsize)
```
Yes, I too think it's not required. I will remove it here.

kotreshhr commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -824,6 +824,338 @@ def test_subvolume_group_info(self):
+        try:
+            self._fs_cmd("subvolumegroup", "resize", self.volname, group, str(nsize), "--no_shrink")
+        except CommandFailedError as ce:
+            self.assertEqual(ce.exitstatus, errno.EINVAL, "invalid error code on resize of subvolumegroup with quota less than used")
```
It is ` usedsize // 2 ` right ?

ajarr replied:
got it! :-)

kotreshhr commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -824,6 +824,338 @@ def test_subvolume_group_info(self):
+
+    def test_subvolume_group_resize_expand_on_full_subvolume(self):
+        """
+        That the subvolume group can be expanded from a full subvolume and future writes succeed.
```
Done

kotreshhr commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -824,6 +824,338 @@ def test_subvolume_group_info(self):
+            time.sleep(15)
+            self.mount_a.write_n_mb(os.path.join(subvolpath, filename), file_size)
+        except CommandFailedError:
+            # Not able to write. So expand the subvolume more and try writing the 5MB file again
```
Done

kotreshhr commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -748,6 +754,76 @@ def test_subvolume_group_create_with_invalid_data_pool_layout(self):
+        for md in group_md:
+            self.assertIn(md, group_info, "'{0}' key not present in metadata of subvolume".format(md))
+
+        self.assertNotEqual(group_info["bytes_pcent"], "undefined", "bytes_pcent should not be set to undefined if quota is not set")
```
Done

kotreshhr commented on src/pybind/mgr/volumes/fs/volume.py:
```diff
@@ -595,13 +597,19 @@ def create_subvolume_group(self, **kwargs):
+                        attrs = {
+                            'uid': uid if uid else 0,
+                            'gid': gid if gid else 0,
+                            'data_pool': pool,
```
I have addressed the issue for subvoluemgroup. PTAL

kotreshhr commented:
> [2a7ac76](https://github.com/ceph/ceph/commit/2a7ac768d37492924dc159a283ed69610bb38179) is the tricky one. So does this commit fix the issue https://tracker.ceph.com/issues/16884 for legacy subvolumes?
> 
> So for legacy and v1 subvolumes whose subvolume groups have quota, the trash dir is to moved /volumes/<subvolgp name>/.trash/ ?

Yes, that's right!

kotreshhr commented on src/pybind/mgr/volumes/fs/volume.py:
```diff
@@ -595,13 +597,21 @@ def create_subvolume_group(self, **kwargs):
+                            'gid': gid if gid is not None else group.gid,
+                            'mode': octal_str_to_decimal_int(mode) if mode else group.mode,
+                            'data_pool': pool,
+                            'quota': size
```
Also, the idempotent subvolumegroup creation might mess up the quota. Imagine the scenario below.

1. Subvoluemgroup is created with quota
2. Idemptoent subvolumegroup creation is done to set only `uid/gid` without passing `quota`

This would reset the `quota` set earlier to `infinite`. This is incorrect and should pick the existing quota right ?
I think this bug exists in subvolume idempotent creation as well.

kotreshhr replied:
Ok, I get it. I thought, it should behave as 'update'. Then we might have to fix the subvolume behavior for the 'uid/gid'.

kotreshhr commented:
> > [2a7ac76](https://github.com/ceph/ceph/commit/2a7ac768d37492924dc159a283ed69610bb38179) is the tricky one. So does this commit fix the issue https://tracker.ceph.com/issues/16884 for legacy subvolumes?
> > So for legacy and v1 subvolumes whose subvolume groups have quota, the trash dir is to moved /volumes//.trash/ ?
> 
> Yes, that's right!

Sorry, it's `/volumes/<group_name>/.trash/`
There is a side effect of doing this. The subvolumegroup can't be removed until the `.trash` is cleaned up which is asynchronous. We need decide how to go about it.

ajarr commented:
> Sorry, it's `/volumes/<group_name>/.trash/`
> There is a side effect of doing this. The subvolumegroup can't be removed until the `.trash` is cleaned up which is asynchronous. We need decide how to go about it.

Good point! So the subvolume group appears empty to the user, and yet he can't unlink it?

ajarr reviewed (commented):
I think we need a test to make sure that legacy subvolumes can be removed when quota is set on their subvolume group.

kotreshhr commented:
> > Sorry, it's `/volumes/<group_name>/.trash/`
> > There is a side effect of doing this. The subvolumegroup can't be removed until the `.trash` is cleaned up which is asynchronous. We need decide how to go about it.
> 
> Good point! So the subvolume group appears empty to the user, and yet he can't unlink it?

Yes, that's correct.

kotreshhr commented on src/pybind/mgr/volumes/module.py:
```diff
@@ -436,8 +437,8 @@ def _cmd_fs_subvolumegroup_create(self, inbuf, cmd):
         """
         return self.vc.create_subvolume_group(
-            vol_name=cmd['vol_name'], group_name=cmd['group_name'],
-            pool_layout=cmd.get('pool_layout', None), mode=cmd.get('mode', '755'),
```
Nope.

kotreshhr commented:
jenkins retest this please

ajarr commented:
@kotreshhr is this good for review? do you plan to make any more changes?

kotreshhr commented:
> @kotreshhr is this good for review? do you plan to make any more changes?

Nope, this is good for final review except for the subvolumegroup removal issue mentioned in [1]. I am waiting on rook team to confirm about the requirement.


[1] https://github.com/ceph/ceph/pull/44347#issuecomment-1046874475

ajarr commented on src/pybind/mgr/volumes/fs/operations/versions/subvolume_base.py:
```diff
@@ -127,6 +127,15 @@ def purgeable(self):
+            group_quota = int(self.fs.getxattr(self.group.path,
+                             'ceph.quota.max_bytes').decode('utf-8'))
+        except cephfs.NoData:
+            group_quota = 0
```
group_quota = None ?

kotreshhr replied:
Done

kotreshhr commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -800,6 +800,105 @@ def test_subvolume_group_info(self):
+        # emulate a v1 subvolume -- in a custom group
+        subvol_path = self._create_v1_subvolume(subvolume, subvol_group=group, has_snapshot=False)
+
+        # this would attempt auto-upgrade on access
```
Done

ajarr reviewed (approved):
Thanks, Kotresh.  The PR looks great to me. The tests are pretty extensive. 

Any idea how we can make resolve_trash() code easier to follow. What comments can we add to make it easier to read?

ajarr commented on src/pybind/mgr/volumes/fs/operations/resolver.py:
```diff
@@ -21,9 +21,20 @@ def resolve(vol_spec, path):
+        groupname = None if parts[2] == Group.NO_GROUP_NAME else parts[2]
+        subvolname = parts[3]
+        return (groupname, subvolname)
+    elif len(parts) == 5:
```
Maybe add some comment like the following:
This is to resolve the trashpath of subvolumes which belong to subvolume groups with quota.  The subvolume trash path has the following syntax 
/\<vol_spec.subvolume_prefix\>/\<groupname\>/.trash/\<subvolname\> ?

kotreshhr replied:
Done

ajarr commented on src/pybind/mgr/volumes/fs/operations/resolver.py:
```diff
@@ -21,9 +21,20 @@ def resolve(vol_spec, path):
-    if len(parts) != 6 or os.path.join(parts[0], parts[1]) != vol_spec.subvolume_prefix or \
-       parts[4] != '.trash':
+
+    if len(parts) == 6:
```
Maybe add some comment like the following:
This is to resolve the trashpath of subvolumes that retain snapshots.  The subvolume trash path has the following syntax 
/\<vol_spec.subvolume_prefix\>/\<groupname\>/\<subvolname\>/.trash/\<subvolUUID\> ?

ajarr reviewed (commented):
Sorry that I keep circling back to the commit to fix the issue with subvolume removal when group quota is set. I find the commit non trivial.

So in the commit message you say "This patch solves the issue by using
the existing per subvolume trash directory introduced
for subvolume removal with retain snapshot usecase."  But isn't the issue solved by using per subvolumegroup trashcan directory ?

kotreshhr commented:
> Sorry that I keep circling back to the commit to fix the issue with subvolume removal when group quota is set. I find the commit non trivial.
> 
> So in the commit message you say "This patch solves the issue by using the existing per subvolume trash directory introduced for subvolume removal with retain snapshot usecase." But isn't the issue solved by using per subvolumegroup trashcan directory ?

I meant the this patch uses the subvolume trash directory framework but the trash location is different. I will rephrase it.

kotreshhr commented:
jenkins test make check

kotreshhr commented:
jenkins test make check arm64

kotreshhr commented:
jenkins render docs

kotreshhr commented:
jenkins test make check

ajarr commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -800,6 +800,98 @@ def test_subvolume_group_info(self):
+        """
+        Tests subvolume removal if it's group quota is set.
+        Validates it doesn't hit https://tracker.ceph.com/issues/16884 as subvolume deletion
+        used to rename the subvolume directory to trash directory which resides outside the
```
Does this still hold true?  AIUI, this issue is not hit by disabling quota for the mgr's libcephfs client, and not by changing the path of the trash directory. Don't we continue to rename the subvolume directory to trash directory that resides outside the subvolume group?

Same comment for tests below.

kotreshhr replied:
Yes, you are right. The comment is no longer valid, I will remove the comment and keep the testcase.

kotreshhr commented on src/pybind/mgr/volumes/module.py:
```diff
@@ -102,6 +102,15 @@ class Module(orchestrator.OrchestratorClientMixin, MgrModule):
             'perm': 'r'
         },
+        {
+            'cmd': 'fs subvolumegroup resize '
```
I had the documentation commit. It got missed midway in some PR refresh. I will find and add it.

kotreshhr commented on src/pybind/mgr/volumes/fs/operations/group.py:
```diff
@@ -88,6 +88,36 @@ def list_snapshots(self):
+        except cephfs.Error as e:
+            raise VolumeException(-e.args[0], e.args[1])
+
+        return {'uid': int(st["uid"]),
```
The data_pool is displayed. The subvolume path makes sense, do you think displaying subvolumgroup path has any use case?

ajarr replied:
yeah, not sure about subvolume group path.  It maybe useful? @vshankar what do you think?

kotreshhr commented on src/pybind/mgr/volumes/fs/operations/group.py:
```diff
@@ -88,6 +88,36 @@ def list_snapshots(self):
+                'bytes_quota': "infinite" if nsize == 0 else nsize,
+                'bytes_used': int(usedbytes),
+                'bytes_pcent': "undefined"
+                if nsize == 0
```
Done

kotreshhr replied:
Done

kotreshhr commented on src/client/Client.cc:
```diff
@@ -11140,7 +11140,7 @@ int Client::statfs(const char *path, struct statvfs *stbuf,
 
   // get_quota_root should always give us something
   // because client quotas are always enabled
```
Done

kotreshhr commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -800,6 +800,307 @@ def test_subvolume_group_info(self):
+
+    def test_subvolume_group_resize_infinite_size_future_writes(self):
+        """
+        That a subvolume group can be resized to an infinite size and the future writes succeed.
```
Modified the test

kotreshhr commented on src/pybind/mgr/volumes/module.py:
```diff
@@ -95,6 +95,13 @@ class Module(orchestrator.OrchestratorClientMixin, MgrModule):
             'perm': 'rw'
         },
+        {
+            'cmd': 'fs subvolumegroup info '
```
Done

kotreshhr commented:
jenkins test make check

ajarr reviewed (commented):
LGTM

ajarr commented:
It's good for qa testing so set the label

kotreshhr commented:
The fs:volumes teuthology results:

http://pulpito.front.sepia.ceph.com/khiremat-2022-04-17_09:43:08-fs:volumes-wip-khiremat-44347-subvolumegroup-quota-8-distro-default-smithi/

@vshankar This has client side changes to make quota optional. Please test this PR along with your batch of PRs.

ajarr commented:
@kotreshhr  In ODF, the clients would mount the subvolume export path using ceph ID with subvolume group path restricted MDS caps. Did you locally check whether in that case subvolume group quota is enforced? In test_volumes, I think the clients mount the entire file system using ceph ID that don't have any path restricted MDS caps.

kotreshhr commented:
The fs:volumes teuthology results:

http://pulpito.front.sepia.ceph.com/khiremat-2022-04-17_09:43:08-fs:volumes-wip-khiremat-44347-subvolumegroup-quota-8-distro-default-smithi/

@vshankar This has client side changes to make quota optional. Please test this PR along with your batch of PRs.

kotreshhr commented:
> @kotreshhr In ODF, the clients would mount the subvolume export path using ceph ID with subvolume group path restricted MDS caps. Did you locally check whether in that case subvolume group quota is enforced? In test_volumes, I think the clients mount the entire file system using ceph ID that don't have any path restricted MDS caps.

This limitation has been discussed initially in the tracker comment https://tracker.ceph.com/issues/53509#note-1 and https://tracker.ceph.com/issues/53509#note-3. I had tested it then while validating the comment. I have not tested it recently.

kotreshhr commented:
> @kotreshhr In ODF, the clients would mount the subvolume export path using ceph ID with subvolume group path restricted MDS caps. Did you locally check whether in that case subvolume group quota is enforced? In test_volumes, I think the clients mount the entire file system using ceph ID that don't have any path restricted MDS caps.

This limitation has been discussed initially in the tracker comment https://tracker.ceph.com/issues/53509#note-1 and https://tracker.ceph.com/issues/53509#note-3. I had tested it then while validating the comment. I have not tested it recently.

ajarr commented:
> > @kotreshhr In ODF, the clients would mount the subvolume export path using ceph ID with subvolume group path restricted MDS caps. Did you locally check whether in that case subvolume group quota is enforced? In test_volumes, I think the clients mount the entire file system using ceph ID that don't have any path restricted MDS caps.
> 
> This limitation has been discussed initially in the tracker comment https://tracker.ceph.com/issues/53509#note-1 and https://tracker.ceph.com/issues/53509#note-3. I had tested it then while validating the comment. I have not tested it recently.

I went through the notes yesterday, which made me think if we actually tested the ODF scenario. I wanted to check whether this scenario was validated.

kotreshhr commented:
> > > @kotreshhr In ODF, the clients would mount the subvolume export path using ceph ID with subvolume group path restricted MDS caps. Did you locally check whether in that case subvolume group quota is enforced? In test_volumes, I think the clients mount the entire file system using ceph ID that don't have any path restricted MDS caps.
> > 
> > 
> > This limitation has been discussed initially in the tracker comment https://tracker.ceph.com/issues/53509#note-1 and https://tracker.ceph.com/issues/53509#note-3. I had tested it then while validating the comment. I have not tested it recently.
> 
> I went through the notes yesterday, which made me think if we actually tested the ODF scenario. I wanted to check whether this scenario was validated.

I validated again with this PR. It behaves as expected. Following are the scenarios.

1. Added `client.node` and `client.node1` users as below restricting to subvolumegroup and subvolume path respectively.
 ```
# cat ./keyring 
[mon.]
	key = AQCtbl5i4AAYMhAAEl7ON9EaSUtbgT4oFhN8mw==
	caps mon = "allow *"
[client.admin]
	key = AQCtbl5iQVWfMxAA7pT3nAD5D9KK2RGn680fhQ==
	caps mds = "allow *"
	caps mgr = "allow *"
	caps mon = "allow *"
	caps osd = "allow *"
[client.fs]
	key = AQDCbl5iQptWLxAAvA0z3cXlcXyXxvFPLT7Giw==
[client.fs_a]
	key = AQDCbl5iA6ZwIRAAwWSXX+iOgMpQH4IfG0XzTw==
[client.node]
	key = AQBscF5iWoODJRAAMRTs4a7IFQurllA2zY5FoA==
	caps mds = "allow rw path=/volumes/csi"
	caps mgr = "allow rw"
	caps mon = "allow r"
	caps osd = "allow rw tag cephfs *=*"
[client.node1]
	key = AQDlc15iFrA8ARAAobAv5v+AP8QuUZAt7W1GDw==
	caps mds = "allow rw path=/volumes/csi/sub_0/8bfc180f-b86b-4ecd-9bab-d46307d9cb13"
	caps mgr = "allow rw"
	caps mon = "allow r"
	caps osd = "allow rw tag cephfs *=*"
 ```
 
 2. Mounted the volumes using `client.node1` on /kmnt as below
 ```
  bin/mount.ceph :/volumes/csi/sub_0/8bfc180f-b86b-4ecd-9bab-d46307d9cb13/ /kmnt1 -o name=node1,conf=/root/sandbox/kotresh-ceph/ceph/build/ceph.conf
 ```
 
 3. Subvoumegroup has quota set to 200 MB
 ```
 # bin/ceph fs subvolumegroup info a csi
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
2022-04-19T14:14:29.939+0530 7ffa67da5700 -1 WARNING: all dangerous and experimental features are enabled.
2022-04-19T14:14:29.947+0530 7ffa67da5700 -1 WARNING: all dangerous and experimental features are enabled.
{
    "atime": "2022-04-19 13:47:51",
    "bytes_pcent": "0",
    "bytes_quota": 200000000,
    "bytes_used": 0,
    "created_at": "2022-04-19 13:47:51",
    "ctime": "2022-04-19 13:48:11",
    "data_pool": "cephfs.a.data",
    "gid": 0,
    "mode": 16877,
    "mon_addrs": [
        "10.76.99.125:40015"
    ],
    "mtime": "2022-04-19 13:48:11",
    "uid": 0
}

 ```
 
 4. Wrote around 1GB waiting 20 sec in between to enforce quota, but quota didn't get enforced as expected.
 ```
 # bin/ceph fs subvolumegroup info a csi
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
2022-04-19T14:14:29.939+0530 7ffa67da5700 -1 WARNING: all dangerous and experimental features are enabled.
2022-04-19T14:14:29.947+0530 7ffa67da5700 -1 WARNING: all dangerous and experimental features are enabled.
{
    "atime": "2022-04-19 13:47:51",
    "bytes_pcent": "524.29",
    "bytes_quota": 200000000,
    "bytes_used": 1048576119,
    "created_at": "2022-04-19 13:47:51",
    "ctime": "2022-04-19 13:48:11",
    "data_pool": "cephfs.a.data",
    "gid": 0,
    "mode": 16877,
    "mon_addrs": [
        "10.76.99.125:40015"
    ],
    "mtime": "2022-04-19 13:48:11",
    "uid": 0
}
 ```
 
 5. Deleted the content. Mounted subvolumepath using the `client.node`
 ```
  bin/mount.ceph :/volumes/csi/sub_0/8bfc180f-b86b-4ecd-9bab-d46307d9cb13/ /kmnt -o name=node,conf=/root/sandbox/kotresh-ceph/ceph/build/ceph.conf
 ```
 
 6. Wrote 200 MB and waited for 20 seconds
 ```
# time for i in {1..200};do dd if=/dev/urandom of=file-1MB-$i status=progress bs=1M count=1;done;sync;sleep 20;

 ```
 
 7. Additional writes failed.
 ```
# time for i in {201..210};do dd if=/dev/urandom of=file-1MB-$i status=progress bs=1M count=1;done;sync;
dd: error writing 'file-1MB-201': Disk quota exceeded
1+0 records in
0+0 records out
0 bytes copied, 0.00738591 s, 0.0 kB/s
dd: error writing 'file-1MB-202': Disk quota exceeded
1+0 records in
0+0 records out
0 bytes copied, 0.00696772 s, 0.0 kB/s
dd: error writing 'file-1MB-203': Disk quota exceeded
1+0 records in
0+0 records out
0 bytes copied, 0.00702509 s, 0.0 kB/s
dd: error writing 'file-1MB-204': Disk quota exceeded
1+0 records in
0+0 records out
0 bytes copied, 0.00683548 s, 0.0 kB/s
dd: error writing 'file-1MB-205': Disk quota exceeded
1+0 records in
0+0 records out
0 bytes copied, 0.00706999 s, 0.0 kB/s
dd: error writing 'file-1MB-206': Disk quota exceeded
1+0 records in
0+0 records out
0 bytes copied, 0.00684821 s, 0.0 kB/s
dd: error writing 'file-1MB-207': Disk quota exceeded
1+0 records in
0+0 records out
0 bytes copied, 0.00694797 s, 0.0 kB/s
dd: error writing 'file-1MB-208': Disk quota exceeded
1+0 records in
0+0 records out
0 bytes copied, 0.00695465 s, 0.0 kB/s
dd: error writing 'file-1MB-209': Disk quota exceeded
1+0 records in
0+0 records out
0 bytes copied, 0.00692966 s, 0.0 kB/s
dd: error writing 'file-1MB-210': Disk quota exceeded
1+0 records in
0+0 records out
0 bytes copied, 0.00696766 s, 0.0 kB/s

real	0m0.098s
user	0m0.000s
 ```

vshankar commented:
> I validated again with this PR. It behaves as expected. Following are the scenarios.

Please add a test case. Rest looks good.

kotreshhr commented:
> > > @kotreshhr In ODF, the clients would mount the subvolume export path using ceph ID with subvolume group path restricted MDS caps. Did you locally check whether in that case subvolume group quota is enforced? In test_volumes, I think the clients mount the entire file system using ceph ID that don't have any path restricted MDS caps.
> > 
> > 
> > This limitation has been discussed initially in the tracker comment https://tracker.ceph.com/issues/53509#note-1 and https://tracker.ceph.com/issues/53509#note-3. I had tested it then while validating the comment. I have not tested it recently.
> 
> I went through the notes yesterday, which made me think if we actually tested the ODF scenario. I wanted to check whether this scenario was validated.

I validated again with this PR. It behaves as expected. Following are the scenarios.

1. Added `client.node` and `client.node1` users as below restricting to subvolumegroup and subvolume path respectively.
 ```
# cat ./keyring 
[mon.]
	key = AQCtbl5i4AAYMhAAEl7ON9EaSUtbgT4oFhN8mw==
	caps mon = "allow *"
[client.admin]
	key = AQCtbl5iQVWfMxAA7pT3nAD5D9KK2RGn680fhQ==
	caps mds = "allow *"
	caps mgr = "allow *"
	caps mon = "allow *"
	caps osd = "allow *"
[client.fs]
	key = AQDCbl5iQptWLxAAvA0z3cXlcXyXxvFPLT7Giw==
[client.fs_a]
	key = AQDCbl5iA6ZwIRAAwWSXX+iOgMpQH4IfG0XzTw==
[client.node]
	key = AQBscF5iWoODJRAAMRTs4a7IFQurllA2zY5FoA==
	caps mds = "allow rw path=/volumes/csi"
	caps mgr = "allow rw"
	caps mon = "allow r"
	caps osd = "allow rw tag cephfs *=*"
[client.node1]
	key = AQDlc15iFrA8ARAAobAv5v+AP8QuUZAt7W1GDw==
	caps mds = "allow rw path=/volumes/csi/sub_0/8bfc180f-b86b-4ecd-9bab-d46307d9cb13"
	caps mgr = "allow rw"
	caps mon = "allow r"
	caps osd = "allow rw tag cephfs *=*"
 ```
 
 2. Mounted the volumes using `client.node1` on /kmnt as below
 ```
  bin/mount.ceph :/volumes/csi/sub_0/8bfc180f-b86b-4ecd-9bab-d46307d9cb13/ /kmnt1 -o name=node1,conf=/root/sandbox/kotresh-ceph/ceph/build/ceph.conf
 ```
 
 3. Subvoumegroup has quota set to 200 MB
 ```
 # bin/ceph fs subvolumegroup info a csi
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
2022-04-19T14:14:29.939+0530 7ffa67da5700 -1 WARNING: all dangerous and experimental features are enabled.
2022-04-19T14:14:29.947+0530 7ffa67da5700 -1 WARNING: all dangerous and experimental features are enabled.
{
    "atime": "2022-04-19 13:47:51",
    "bytes_pcent": "0",
    "bytes_quota": 200000000,
    "bytes_used": 0,
    "created_at": "2022-04-19 13:47:51",
    "ctime": "2022-04-19 13:48:11",
    "data_pool": "cephfs.a.data",
    "gid": 0,
    "mode": 16877,
    "mon_addrs": [
        "10.76.99.125:40015"
    ],
    "mtime": "2022-04-19 13:48:11",
    "uid": 0
}

 ```
 
 4. Wrote around 1GB waiting 20 sec in between to enforce quota, but quota didn't get enforced as expected.
 ```
 # bin/ceph fs subvolumegroup info a csi
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
2022-04-19T14:14:29.939+0530 7ffa67da5700 -1 WARNING: all dangerous and experimental features are enabled.
2022-04-19T14:14:29.947+0530 7ffa67da5700 -1 WARNING: all dangerous and experimental features are enabled.
{
    "atime": "2022-04-19 13:47:51",
    "bytes_pcent": "524.29",
    "bytes_quota": 200000000,
    "bytes_used": 1048576119,
    "created_at": "2022-04-19 13:47:51",
    "ctime": "2022-04-19 13:48:11",
    "data_pool": "cephfs.a.data",
    "gid": 0,
    "mode": 16877,
    "mon_addrs": [
        "10.76.99.125:40015"
    ],
    "mtime": "2022-04-19 13:48:11",
    "uid": 0
}
 ```
 
 5. Deleted the content. Mounted subvolumepath using the `client.node`
 ```
  bin/mount.ceph :/volumes/csi/sub_0/8bfc180f-b86b-4ecd-9bab-d46307d9cb13/ /kmnt -o name=node,conf=/root/sandbox/kotresh-ceph/ceph/build/ceph.conf
 ```
 
 6. Wrote 200 MB and waited for 20 seconds
 ```
# time for i in {1..200};do dd if=/dev/urandom of=file-1MB-$i status=progress bs=1M count=1;done;sync;sleep 20;

 ```
 
 7. Additional writes failed.
 ```
# time for i in {201..210};do dd if=/dev/urandom of=file-1MB-$i status=progress bs=1M count=1;done;sync;
dd: error writing 'file-1MB-201': Disk quota exceeded
1+0 records in
0+0 records out
0 bytes copied, 0.00738591 s, 0.0 kB/s
dd: error writing 'file-1MB-202': Disk quota exceeded
1+0 records in
0+0 records out
0 bytes copied, 0.00696772 s, 0.0 kB/s
dd: error writing 'file-1MB-203': Disk quota exceeded
1+0 records in
0+0 records out
0 bytes copied, 0.00702509 s, 0.0 kB/s
dd: error writing 'file-1MB-204': Disk quota exceeded
1+0 records in
0+0 records out
0 bytes copied, 0.00683548 s, 0.0 kB/s
dd: error writing 'file-1MB-205': Disk quota exceeded
1+0 records in
0+0 records out
0 bytes copied, 0.00706999 s, 0.0 kB/s
dd: error writing 'file-1MB-206': Disk quota exceeded
1+0 records in
0+0 records out
0 bytes copied, 0.00684821 s, 0.0 kB/s
dd: error writing 'file-1MB-207': Disk quota exceeded
1+0 records in
0+0 records out
0 bytes copied, 0.00694797 s, 0.0 kB/s
dd: error writing 'file-1MB-208': Disk quota exceeded
1+0 records in
0+0 records out
0 bytes copied, 0.00695465 s, 0.0 kB/s
dd: error writing 'file-1MB-209': Disk quota exceeded
1+0 records in
0+0 records out
0 bytes copied, 0.00692966 s, 0.0 kB/s
dd: error writing 'file-1MB-210': Disk quota exceeded
1+0 records in
0+0 records out
0 bytes copied, 0.00696766 s, 0.0 kB/s

real	0m0.098s
user	0m0.000s
 ```

> > I validated again with this PR. It behaves as expected. Following are the scenarios.
> 
> Please add a test case. Rest looks good.

Sure, I will add it.

kotreshhr commented:
> > I validated again with this PR. It behaves as expected. Following are the scenarios.
> 
> Please add a test case. Rest looks good.

@vshankar @ajarr Added the tests. Please take a look.

kotreshhr commented:
Rebased

kotreshhr commented:
@vshankar PTAL

vshankar commented:
> @vshankar PTAL

In my test branch. Tests running here - https://pulpito.ceph.com/vshankar-2022-04-27_15:24:52-fs-wip-vshankar-testing1-20220427-145314-testing-default-smithi/

vshankar reviewed (changes requested):
@kotreshhr - https://pulpito.ceph.com/vshankar-2022-05-01_13:18:44-fs-wip-vshankar-testing1-20220428-204527-testing-default-smithi/6817529/

```
022-05-01T14:47:38.939 INFO:tasks.cephfs_test_runner:======================================================================
2022-05-01T14:47:38.940 INFO:tasks.cephfs_test_runner:ERROR: test_subvolume_group_quota_mds_path_restriction_to_group_path (tasks.cephfs.test_volumes.TestSubvolumeGroups)
2022-05-01T14:47:38.940 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
2022-05-01T14:47:38.940 INFO:tasks.cephfs_test_runner:Traceback (most recent call last):
2022-05-01T14:47:38.940 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_b82312f270e887c49dafe642c4ee39a73bc7d8fa/qa/tasks/cephfs/test_volumes.py", line 957, in test_subvolume_group_quota_mds_path_restriction_to_group_path
2022-05-01T14:47:38.940 INFO:tasks.cephfs_test_runner:    self.mgr_cluster.admin_remote.run(args=args, wait=True)
2022-05-01T14:47:38.940 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_git_teuthology_fd5c31669e51177d3c57bc0420279c5d1e944497/teuthology/orchestra/remote.py", line 510, in run
2022-05-01T14:47:38.940 INFO:tasks.cephfs_test_runner:    r = self._runner(client=self.ssh, name=self.shortname, **kwargs)
2022-05-01T14:47:38.940 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_git_teuthology_fd5c31669e51177d3c57bc0420279c5d1e944497/teuthology/orchestra/run.py", line 455, in run
2022-05-01T14:47:38.940 INFO:tasks.cephfs_test_runner:    r.wait()
2022-05-01T14:47:38.940 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_git_teuthology_fd5c31669e51177d3c57bc0420279c5d1e944497/teuthology/orchestra/run.py", line 161, in wait
2022-05-01T14:47:38.941 INFO:tasks.cephfs_test_runner:    self._raise_for_status()
2022-05-01T14:47:38.941 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_git_teuthology_fd5c31669e51177d3c57bc0420279c5d1e944497/teuthology/orchestra/run.py", line 183, in _raise_for_status
2022-05-01T14:47:38.941 INFO:tasks.cephfs_test_runner:    node=self.hostname, label=self.label
2022-05-01T14:47:38.941 INFO:tasks.cephfs_test_runner:teuthology.exceptions.CommandFailedError: Command failed on smithi136 with status 1: 'ceph-authtool /etc/ceph/ceph.client.guest1.keyring --import-keyring /tmp/client.guest1.keyring'
2022-05-01T14:47:38.941 INFO:tasks.cephfs_test_runner:
2022-05-01T14:47:38.941 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
```

kotreshhr commented:
Teuthology Results of fs:volumes suite:
http://pulpito.front.sepia.ceph.com/khiremat-2022-05-06_12:49:38-fs:volumes-wip-khiremat-44347-subvolumegroup-quota-final-3-distro-default-smithi/

kotreshhr commented:
jenkins test api

vshankar commented:
@kotreshhr fails again - https://pulpito.ceph.com/vshankar-2022-05-13_17:09:16-fs-wip-vshankar-testing-20220513-120051-testing-default-smithi/6832872

```
2022-05-13T18:21:20.628 INFO:tasks.cephfs_test_runner:======================================================================
2022-05-13T18:21:20.628 INFO:tasks.cephfs_test_runner:FAIL: test_subvolume_group_quota_mds_path_restriction_to_group_path (tasks.cephfs.test_volumes.TestSubvolumeGroups)
2022-05-13T18:21:20.628 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
2022-05-13T18:21:20.628 INFO:tasks.cephfs_test_runner:Traceback (most recent call last):
2022-05-13T18:21:20.628 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_0aa621207192df0365b7a68dd0219f85313a7767/qa/tasks/cephfs/test_volumes.py", line 979, in test_subvolume_group_quota_mds_path_restriction_to_group_path
2022-05-13T18:21:20.629 INFO:tasks.cephfs_test_runner:    self.fail(f"expected filling subvolume {subvolname} with 400 files of size 1MB to fail")
2022-05-13T18:21:20.629 INFO:tasks.cephfs_test_runner:AssertionError: expected filling subvolume subvolume_0000000000802628 with 400 files of size 1MB to fail
2022-05-13T18:21:20.629 INFO:tasks.cephfs_test_runner:
2022-05-13T18:21:20.629 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
```

Looks like a quote enforcement delay? If that's the case please run this change through fs:volumes (client/mds tests look good). I'm dropping this changes from my current batch.

vshankar reviewed (changes requested):
PTAL at the failure.

kotreshhr commented:
> @kotreshhr fails again - https://pulpito.ceph.com/vshankar-2022-05-13_17:09:16-fs-wip-vshankar-testing-20220513-120051-testing-default-smithi/6832872
> 
> ```
> 2022-05-13T18:21:20.628 INFO:tasks.cephfs_test_runner:======================================================================
> 2022-05-13T18:21:20.628 INFO:tasks.cephfs_test_runner:FAIL: test_subvolume_group_quota_mds_path_restriction_to_group_path (tasks.cephfs.test_volumes.TestSubvolumeGroups)
> 2022-05-13T18:21:20.628 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
> 2022-05-13T18:21:20.628 INFO:tasks.cephfs_test_runner:Traceback (most recent call last):
> 2022-05-13T18:21:20.628 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_0aa621207192df0365b7a68dd0219f85313a7767/qa/tasks/cephfs/test_volumes.py", line 979, in test_subvolume_group_quota_mds_path_restriction_to_group_path
> 2022-05-13T18:21:20.629 INFO:tasks.cephfs_test_runner:    self.fail(f"expected filling subvolume {subvolname} with 400 files of size 1MB to fail")
> 2022-05-13T18:21:20.629 INFO:tasks.cephfs_test_runner:AssertionError: expected filling subvolume subvolume_0000000000802628 with 400 files of size 1MB to fail
> 2022-05-13T18:21:20.629 INFO:tasks.cephfs_test_runner:
> 2022-05-13T18:21:20.629 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
> ```
> 
> Looks like a quote enforcement delay? If that's the case please run this change through fs:volumes (client/mds tests look good). I'm dropping this changes from my current batch.

Yes, it is. I have tweaked a testcase a bit to include another delay after the quota is exceeded. Let's see. I will trigger fs:volumes suite and attach the results.

nmshelke commented:
@kotreshhr Can we add subvolume count as a part of `ceph fs subvolumegroup info` command?
@Madhu-1 Feel free to add your doubts/requirements.

kotreshhr commented:
> @kotreshhr Can we add subvolume count as a part of `ceph fs subvolumegroup info` command? @Madhu-1 Feel free to add your doubts/requirements.
@Madhu-1 what's the usecase ? The `mgr/volumes` would also need to do`ls` on the subvolumegroup and count. I have not tested but this might slow down info command on subvolumegroup with thousands of subvolumes.

Madhu-1 commented:
> > @kotreshhr Can we add subvolume count as a part of `ceph fs subvolumegroup info` command? @Madhu-1 Feel free to add your doubts/requirements.
> 
> @Madhu-1 what's the usecase ? The `mgr/volumes` would also need to do`ls` on the subvolumegroup and count. I have not tested but this might slow down info command on subvolumegroup with thousands of subvolumes.

@kotreshhr use-case mentioned here https://github.com/ceph/ceph/pull/45949/files#r881728391. in scale setup listing all subvolumes to check group is empty or not is a slow operation. it will affect the performance as you mentioned. 

Another question same is applicable for volume info also? if we have many subvolumegroups it will also slow down the info command?

kotreshhr commented:
> > > @kotreshhr Can we add subvolume count as a part of `ceph fs subvolumegroup info` command? @Madhu-1 Feel free to add your doubts/requirements.
> > 
> > 
> > @Madhu-1 what's the usecase ? The `mgr/volumes` would also need to do`ls` on the subvolumegroup and count. I have not tested but this might slow down info command on subvolumegroup with thousands of subvolumes.
> 
> @kotreshhr use-case mentioned here https://github.com/ceph/ceph/pull/45949/files#r881728391. in scale setup listing all subvolumes to check group is empty or not is a slow operation. it will affect the performance as you mentioned.
> 
> Another question same is applicable for volume info also? if we have many subvolumegroups it will also slow down the info command?

Yes, I don't see any difference between 'mgr/volumes' doing it or the application doing it. Performance wise, it's all the same.

kotreshhr commented:
> > @kotreshhr fails again - https://pulpito.ceph.com/vshankar-2022-05-13_17:09:16-fs-wip-vshankar-testing-20220513-120051-testing-default-smithi/6832872
> > ```
> > 2022-05-13T18:21:20.628 INFO:tasks.cephfs_test_runner:======================================================================
> > 2022-05-13T18:21:20.628 INFO:tasks.cephfs_test_runner:FAIL: test_subvolume_group_quota_mds_path_restriction_to_group_path (tasks.cephfs.test_volumes.TestSubvolumeGroups)
> > 2022-05-13T18:21:20.628 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
> > 2022-05-13T18:21:20.628 INFO:tasks.cephfs_test_runner:Traceback (most recent call last):
> > 2022-05-13T18:21:20.628 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_0aa621207192df0365b7a68dd0219f85313a7767/qa/tasks/cephfs/test_volumes.py", line 979, in test_subvolume_group_quota_mds_path_restriction_to_group_path
> > 2022-05-13T18:21:20.629 INFO:tasks.cephfs_test_runner:    self.fail(f"expected filling subvolume {subvolname} with 400 files of size 1MB to fail")
> > 2022-05-13T18:21:20.629 INFO:tasks.cephfs_test_runner:AssertionError: expected filling subvolume subvolume_0000000000802628 with 400 files of size 1MB to fail
> > 2022-05-13T18:21:20.629 INFO:tasks.cephfs_test_runner:
> > 2022-05-13T18:21:20.629 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Looks like a quote enforcement delay? If that's the case please run this change through fs:volumes (client/mds tests look good). I'm dropping this changes from my current batch.
> 
> Yes, it is. I have tweaked the testcases a bit to include another delay after the quota is exceeded. Let's see. I will trigger fs:volumes suite and attach the results.

@vshankar I found that the test case passes some times and fails most of the times. I found only distro and mount style to be the only difference.
Passed one args: http://qa-proxy.ceph.com/teuthology/khiremat-2022-05-06_12:49:38-fs:volumes-wip-khiremat-44347-subvolumegroup-quota-final-3-distro-default-smithi/6824949/info.yaml
```
description: fs:volumes/{begin/{0-install 1-ceph 2-logrotate} clusters/1a3s-mds-4c-client
  conf/{client mds mon osd} distro/{ubuntu_latest} mount/kclient/{mount-syntax/{v2}
  mount overrides/{distro/testing/k-testing ms-die-on-skipped}} objectstore/bluestore-bitmap
  overrides/{no_client_pidfile whitelist_health whitelist_wrongly_marked_down} tasks/volumes/{overrides
  test/basic}}
```

Failed one args: http://qa-proxy.ceph.com/teuthology/khiremat-2022-05-17_12:30:55-fs:volumes-wip-khiremat-44347-subvolumegroup-quota-final-4-distro-default-smithi/6838373/info.yaml
```
description: fs:volumes/{begin/{0-install 1-ceph 2-logrotate} clusters/1a3s-mds-4c-client
  conf/{client mds mon osd} distro/{centos_8} mount/kclient/{mount-syntax/{v1} mount
  overrides/{distro/testing/k-testing ms-die-on-skipped}} objectstore/bluestore-bitmap
  overrides/{no_client_pidfile whitelist_health whitelist_wrongly_marked_down} tasks/volumes/{overrides
  test/basic}}
```

vshankar commented:
> ount/kclient/{mount-syntax/{v1} mount

Checking this now...

vshankar commented:
@kotreshhr I did some yaml hacking to run the failed tests with v1 and v2 syntax with centos_8

- https://pulpito.ceph.com/vshankar-2022-05-27_18:25:32-fs:volumes-wip-vshankar-testing-20220513-120051-testing-default-smithi/6850780/ (v1 syntax)
- https://pulpito.ceph.com/vshankar-2022-05-27_07:49:07-fs:volumes-wip-vshankar-testing-20220513-120051-testing-default-smithi/6850607/ (v2 syntax)

Both test failed with the same failure as seen earlier. I think it *might* be related to the specific kernel being used.

kotreshhr commented:
> @kotreshhr I did some yaml hacking to run the failed tests with v1 and v2 syntax with centos_8
> 
>     * https://pulpito.ceph.com/vshankar-2022-05-27_18:25:32-fs:volumes-wip-vshankar-testing-20220513-120051-testing-default-smithi/6850780/ (v1 syntax)
> 
>     * https://pulpito.ceph.com/vshankar-2022-05-27_07:49:07-fs:volumes-wip-vshankar-testing-20220513-120051-testing-default-smithi/6850607/ (v2 syntax)
> 
> 
> Both test failed with the same failure as seen earlier. I think it _might_ be related to the specific kernel being used.

Yes, you are right. It's something to do with lsm modules which fills  inodes' `i_security` field. Please find the detail in this tracker [1] explained. As a resolution, we need to give `read access` to the parent of the directory on which quota is being set.

In this particular subvolumegroup use case, `r /volumes, rw /volumes/csi/<subvol>/<uuid>/` or `rw /volumes/` should do the trick. 

I have reproduced this issue locally and tested the workaround. It works. I will refresh the PR with the following changes.

1. Change qa test giving read access to `/volumes`
2. Update quota limitation doc to reflect the above.

@vshankar Does this sound good ?

[1] https://tracker.ceph.com/issues/55090#note-8

kotreshhr commented:
jenkins test make check

kotreshhr commented:
Teuthology Run Scheduled:
http://pulpito.front.sepia.ceph.com/khiremat-2022-06-02_03:34:59-fs:volumes-wip-khiremat-44347-subvolumegroup-quota-0000-distro-default-smithi/

vshankar commented:
Nice work @kotreshhr

vshankar commented on doc/cephfs/quota.rst:
```diff
@@ -41,6 +41,13 @@ Limitations
+   enforce them. If quota is configured on a directory path
+   (e.g., ``/home/volumes/group``), the kclient needs to have access
+   to the parent (e.g., ``/home/volumes``).
+   See also: https://tracker.ceph.com/issues/55090
```
Should we include an example command on how to create the user?

kotreshhr replied:
We can do that.

kotreshhr replied:
I will add the following example command.
```
ceph auth get-or-create client.guest1 mds 'allow r path=/home/volumes, allow rw path=/home/volumes/group' mgr 'allow rw' osd 'allow rw tag cephfs metadata=*' mon 'allow r'
```

kotreshhr replied:
@vshankar Done

vshankar commented:
@kotreshhr Tests run looks good :)

vshankar reviewed (approved):
https://tracker.ceph.com/projects/cephfs/wiki/Master#2022-Jun-13
